Add flag to set CNI bin dir, and use it on gci nodes

bboreham opened the pull request:
**What this PR does / why we need it**:

When using `kube-up` on GCE, following #31023 which moved the workers from debian to gci, CNI just isn't working.  The root cause is basically as discussed in #28563: one flag (`--network-plugin-dir`) means two different things, and the `configure-helper` script uses it for the wrong purpose.

This PR adds a new flag `--cni-bin-dir`, then uses it to configure CNI as desired.

As discussed at #28563, I have also added a flag `--cni-conf-dir` so users can be explicit 

**Which issue this PR fixes** : fixes #28563

**Special notes for your reviewer**:

I left the old flag largely alone for backwards-compatibility, with the exception that I stop setting the default when CNI is in use.  The value of `"/usr/libexec/kubernetes/kubelet-plugins/net/exec/"` is unlikely to be what is wanted there.

**Release note**:

``` release-note
Added new kubelet flags `--cni-bin-dir` and `--cni-conf-dir` to specify where CNI files are located.
Fixed CNI configuration on GCI platform when using CNI.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32151)

<!-- Reviewable:end -->

bboreham committed:
Add options to set CNI config and binary directories

Also make clearer the function of --network-plugin-dir when using CNI

bboreham committed:
Use --cni-bin-dir flag when using CNI on GCI

bboreham commented:
Updated and rebased.  @freehan this is the PR I mentioned in #27027.

bboreham commented:
Oh, one question: the default for NetworkPluginDir is in `pkg/apis/componentconfig/v1alpha1/defaults.go` but the defaults for CNI dirs are in `pkg/kubelet/network/cni/cni.go`.  Is it desirable to bring those into alignment?

errordeveloper commented on cmd/kubelet/app/options/options.go:
```diff
@@ -147,7 +147,9 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&s.VolumeStatsAggPeriod.Duration, "volume-stats-agg-period", s.VolumeStatsAggPeriod.Duration, "Specifies interval for kubelet to calculate and cache the volume disk usage for all pods and volumes.  To disable volume calculations, set to 0.  Default: '1m'")
 	fs.StringVar(&s.NetworkPluginName, "network-plugin", s.NetworkPluginName, "<Warning: Alpha feature> The name of the network plugin to be invoked for various events in kubelet/pod lifecycle")
-	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins")
+	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins or CNI config")
```
You probably want to mark it deprecated...

bboreham replied:
That would only be appropriate for the CNI usage of the flag; there is another meaning for non-CNI plugins.
Maybe non-CNI plugins should be deprecated, but that's another conversation.

errordeveloper commented:
I think you also need to run `make generated_files`, as you have updated `types.go`.

bboreham commented:
> I think you also need to run make generated_files, as you have updated types.go

OK, have run that and checked in the resulting changes.  Thanks!

caseydavenport commented on cmd/kubelet/app/options/options.go:
```diff
@@ -147,7 +147,9 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.NetworkPluginName, "network-plugin", s.NetworkPluginName, "<Warning: Alpha feature> The name of the network plugin to be invoked for various events in kubelet/pod lifecycle")
-	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins")
+	fs.StringVar(&s.NetworkPluginDir, "network-plugin-dir", s.NetworkPluginDir, "<Warning: Alpha feature> The full path of the directory in which to search for network plugins or CNI config")
+	fs.StringVar(&s.CNIConfDir, "cni-conf-dir", s.CNIConfDir, "<Warning: Alpha feature> The full path of the directory in which to search for CNI config files")
```
Could we include the default values for these options?  Most of the other options have them listed.

bboreham replied:
Sure; any views on my question about which file the default should be implemented in?

dcbw commented:
> Oh, one question: the default for NetworkPluginDir is in pkg/apis/componentconfig/v1alpha1/defaults.go but the defaults for CNI dirs are in pkg/kubelet/network/cni/cni.go. Is it desirable to bring those into alignment?

I feel like the defaults should be left up to each plugin type, either exec or CNI.  I think that was just an omission in Rajat's original CNI branch.  I don't think a kubelet global default is really appropriate here, since different plugins have different conventions.  So maybe remove the default from defaults.go and put it into exec.go?

bboreham committed:
Move default directory for exec plugin into exec plugin

bboreham commented:
I added the default values to the usage messages and moved the 'exec' plugin default directory into `exec.go`.  Also squashed commits a bit.

> I think we can do CNI defaults in the Flags() calls

This doesn't seem to be the convention in `kubernetes/cmd/kubelet/app/options/options.go`.  All the flags there are using the current value from the struct (invariably the zero value) as a default, and those that display a default to the user have it hard-coded in the usage string.

dcbw commented:
LGTM

bboreham commented:
Could someone mark this as OK to test, please?  I'm quite keen to see CNI working on the default `kube-up` configuration.

luxas commented:
ok to test

caseydavenport commented:
@bboreham Thanks once again for doing the legwork for this! 

Can we please get this added to the v1.4 milestone, or get https://github.com/kubernetes/kubernetes/pull/31023 reverted?  This breaks CNI on GCE deployments...

freehan commented:
Please fix the cni network plugin unit tests
```
FAIL	k8s.io/kubernetes/pkg/kubelet/network/cni [build failed]
```

Also check the error from `verify-flags-underscore.py`. Probably need to add new flag into `https://github.com/kubernetes/kubernetes/blob/master/hack/verify-flags/known-flags.txt`
```
FAILED   hack/make-rules/../../hack/verify-flags-underscore.py	3s
```

thockin commented:
@pwittrock

thockin commented:
Ugh.  Just to be clear, this is bad, but maybe not "force it into 1.4.0" bad.  I think this could wait for 1.4.1 to integrate.

thockin commented:
I dislike how much of the individual net plugins are being exposed by flags now.  If we take a step back, could we define a single network-plugin flag that carried key-value params from flags to drivers?

bboreham commented:
@thockin to be clear, you are reversing the position at https://github.com/kubernetes/kubernetes/issues/28563#issuecomment-230983339, where you wanted CNI in the name?

freehan commented:
@bboreham do you need help to fix the tests?

bboreham commented:
@freehan thanks for the offer; I think I am on top of it now.

bboreham committed:
Fix unit tests and known flags list

vishh commented:
I talked to @thockin offline and he is OK with this PR as-is. @freehan let's get this PR merged asap and have it cherry-picked into the 1.4 release branch.

freehan commented:
Looks like the last missing piece:


```
FAILED   hack/make-rules/../../hack/verify-codecgen.sh	731s
```
Rerun hack/update-codecgen.sh

freehan commented:
I think you have to run `hack/update-codecgen.sh` once again.

bboreham committed:
Regenerate codecgen files

freehan commented:
Thanks. LGTM. assuming tests will pass.
